File: src/frontend/parser/programParser.ts
// imports 
import { tokenize } from "../lexer/tokenize";
import ParserContext from "./parserContext";
import StatementParser from "./statementParser";
import { Program, ProgramKind } from "./types";

export default class ProgramParser {

  // an AST is built by parsing program 
  buildAST(sourceCode: string): Program {
    const program: Program = {
      kind: ProgramKind.Program,
      body: [],
    };
    
    const tokens = tokenize(sourceCode); 
    const parserContext = new ParserContext(tokens, 0); 
    const statementParser = new StatementParser(parserContext);

    while(!parserContext.endOf()){
      const stmt = statementParser.parseStatement();
      program.body.push(stmt);
    }

    return program;
  }
}

File: src/frontend/parser/program_parser.cpp
#include <scriptum/frontend/lexer/token.hpp>
#include <scriptum/frontend/parser/program_parser.hpp>
#include <scriptum/frontend/parser/expression_parser.hpp>
#include <scriptum/frontend/parser/parser_context.hpp>
#include <scriptum/frontend/parser/statement_parser.hpp>

namespace scriptum {

Program ProgramParser::buildAST(const std::string& sourceCode)
{
    Program program;

    const std::vector<Token> tokens = tokenize(sourceCode);
    ParserContext parserContext(tokens, 0);
    ExpressionParser expressionParser(parserContext);
    StatementParser statementParser(parserContext, expressionParser);

    while (!parserContext.endOf())
    {
        program.body.push_back(statementParser.parseStatement());
    }

    return program;
}

} // namespace scriptum
